### PR TITLE
Shapes API refractor #779 and #644

### PR DESCRIPTION
**Pull request by yaustar**
pc.Plane, pc.BoundingBox and pc.BoundingSphere now have a standardised function for intersectsRay and all are capable of calculating the intersection point.

Added documentation for intersectsRay.

Constructors now clone the params that are passed rather than store a reference.

Removed any allocations of pc.Vec3 in the functions.

### CODE REVIEW

File: src/shape/plane.js [hunk 1]
```diff
@@ -1,36 +1,96 @@
...
         },
 
-        intersect: function (p0, p1) {
-            var d0 = this.distance(p0);
-            var d1 = this.distance(p1);
+        /**
+         * @function
+         * @name pc.Plane#intersectsLine
+         * @description Test if the plane intersects between two points.
+         * @param {pc.Vec3} p0 First point position.
```
Comment (daredevildave) [inline src/shape/plane.js]: start, end?

File: src/shape/plane.js [hunk 2]
```diff
@@ -1,34 +1,64 @@
...
-        this.point  = point || new pc.Vec3(0, 0, 0);
-        this.d = -this.normal.dot(this.point);
+        this.normal = normal ? normal.clone() : new pc.Vec3(0, 0, 1);
+        this.point = point ? point.clone() : new pc.Vec3(0, 0, 0);
     };
 
     Plane.prototype = {
-        distance: function (pos) {
-            return this.normal.dot(pos) + this.d;
+        _intersect: function (p0, p1) {
```
Comment (yaustar) [inline src/shape/plane.js]: Could this be inlined?

**Comment by yaustar** [comment]
Still needs testing on other projects

**Comment by yaustar** [comment]
Tested against other projects and looks good

File: src/shape/bounding-sphere.js [hunk 3]
```diff
@@ -81,16 +89,22 @@ pc.extend(pc, function () {
...
-            return ray.direction.clone().scale(t).add(ray.origin);
+            if (point)
+                point.copy(ray.direction).scale(t).add(ray.origin);
+
+            return true;
         }
     };
 
+    // For backwards compatibility
+    BoundingSphere.prototype.intersectRay = BoundingSphere.prototype.intersectsRay;
```
Comment (yaustar) [inline src/shape/bounding-sphere.js]: Done and useful to know

**Pull request merged**